File: julia/permutations.jl
function printpermutations(a, n)
    if n <= 0
        println(join(a, "\t"))
    else
        for i = 1:n-1
            printpermutations(a, n-1)
            local j = iseven(n) ? i : 1
            a[j], a[n] = a[n], a[j]
        end
        printpermutations(a, n-1)
    end
end

printpermutations(ARGS, length(ARGS))

File: nim/permutations.nim
import std/[strutils, cmdline]

type InputArgumentError* = object of IOError

proc printPermutations(a: var seq[string], n: int) =
  if n <= 0:
    echo a.join("\t")
  else:
    for i in 0 ..< n:
      printPermutations(a, n - 1)
      let j = if n mod 2 == 0: 0 else: i
      swap(a[j], a[n])
    printPermutations(a, n - 1)

when isMainModule:
  var args = commandLineParams()
  if len(args) == 0:
    raise newException(InputArgumentError,
      "ERROR: At least one argument is required")
  printPermutations(args, len(args) - 1)
